src/ott/solvers/linear/univariate.py

```diff
@@ -12,134 +12,247 @@
+    x, y = geom.x, geom.y
+
+    # check if problem has the property uniform / same number of points
+    is_uniform_same_size = prob.is_uniform and prob.is_equal_size
+    if self.num_subsamples:
+      rng1, rng2 = jax.random.split(rng, 2)
+      if prob.is_uniform:
+        x = x[jnp.linspace(0, n, num=self.num_subsamples).astype(int), :]
+        y = y[jnp.linspace(0, m, num=self.num_subsamples).astype(int), :]
+      else:
+        x = jax.random.choice(rng1, x, (self.num_subsamples,), p=prob.a, axis=0)
+        y = jax.random.choice(rng2, y, (self.num_subsamples,), p=prob.b, axis=0)
+        n = m = self.num_subsamples
+      # now that both are subsampled, consider them as uniform/same size.
+      is_uniform_same_size = True
+
+    if self.quantiles is None:
+      if is_uniform_same_size:
+        i_x, i_y = jnp.argsort(x, axis=0), jnp.argsort(y, axis=0)
+        x = jnp.take_along_axis(x, i_x, axis=0)
+        y = jnp.take_along_axis(y, i_y, axis=0)
+        ot_costs = jax.vmap(geom.cost_fn.h, in_axes=[0])(x.T - y.T) / n
+
+        if self.num_subsamples:
+          # When subsampling, the pairing computed have no meaning w.r.t.
+          # original data.
+          paired_indices, mass_paired_indices = None, None
+        else:
+          paired_indices = jnp.stack([i_x, i_y]).transpose([2, 0, 1])
+          mass_paired_indices = jnp.ones((n,)) / n
+
+      else:
+        ot_costs, paired_indices, mass_paired_indices = jax.vmap(
+            self._quantile_distance_and_transport,
+            in_axes=[1, 1, None, None, None]
+        )(x, y, prob.a, prob.b, geom.cost_fn)
+
     else:
-      raise NotImplementedError(f"Method `{self.method}` not implemented.")
+      assert prob.is_uniform, "Quantile method only valid for uniform marginals"
+      x_q = jnp.quantile(x, self.quantiles, axis=0)
+      y_q = jnp.quantile(y, self.quantiles, axis=0)
+      ot_costs = jax.vmap(geom.cost_fn.pairwise, in_axes=[1, 1])(x_q, y_q)
+      ot_costs /= self.num_quantiles
+      paired_indices = None
+      mass_paired_indices = None
 
-    # re-scale when subsampling
-    return self.cost_fn.pairwise(xx, yy) * (n / xx.shape[0])
+    return UnivariateOutput(
+        prob=prob,
+        ot_costs=ot_costs,
+        paired_indices=paired_indices,
+        mass_paired_indices=mass_paired_indices
+    )
 
-  def _cdf_distance(
-      self, x: jnp.ndarray, y: jnp.ndarray, a: Optional[jnp.ndarray],
-      b: Optional[jnp.ndarray]
+  def _quantile_distance_and_transport(
+      self, x: jnp.ndarray, y: jnp.ndarray, a: jnp.ndarray, b: jnp.ndarray,
+      cost_fn: costs.TICost
   ):
-    # Implementation based on `scipy` implementation for
+    # Implementation inspired by `scipy` implementation for
     # :func:<scipy.stats.wasserstein_distance>
-    a = jnp.ones_like(x) if a is None else a
-    a /= jnp.sum(a)
-    b = jnp.ones_like(y) if b is None else b
-    b /= jnp.sum(b)
+    def sort_and_argsort(x: jnp.array, return_argsort: bool, **kwargs):
```

Comment:
Think the `kwargs` can be removed.

Reply:
Would also use `def sort_and_argsort(x: jnp.array, *, argsort: bool) -> <add correct type hint>`

Reply:
thanks, great suggestions. moved it to ``maths.utils.py`` and followed your suggestions